### PR TITLE
[WIP] Field-aware factorization machines

### PR DESCRIPTION
**Pull request by mdymczyk**
Initial implementation of field-aware factorization machines.

Based on these 2 whitepapers:
* https://arxiv.org/pdf/1701.04099.pdf
* https://www.csie.ntu.edu.tw/~cjlin/papers/ffm.pdf

And the following repositories:
* https://github.com/guestwalk/libffm (original impl)
* https://github.com/alexeygrigorev/libffm-python (Python interfact for it)
* https://github.com/RTBHOUSE/cuda-ffm (CUDA implementation of a simplified method)

Currently only initial GPU implementation as CPU will most probably just be a copy of the original impl (without the SSE alignments for now).

No benchmarks for far as there's still something wrong (getting different results).

Thing to be still done:

* add validation set option and early stopping (FFM seems to need this a lot as it tends to overfit)
* add multi GPU support
* review the data structures used - using an object oriented approach with Dataset/Row/Node hierarchy is good for development but might provide a lot of overhead when copying data to the device, refactoring this into 3 (or more) continuous arrays might provide a lot of speedup
* review the main method `wTx` (in `trainer.cu`) - probably can be rewritten in a more GPU friendly manner
* probably something else I'm forgetting

### CODE REVIEW

File: src/gpu/ffm/trainer.cu [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+  int n1 = rowSizes[rowIdx] + nodeIdx - cBatchOffset[0];
+
+  fieldFeature[threadIdx.x * 2] = fields[n1];
+  fieldFeature[threadIdx.x * 2 + 1] = features[n1];
+  vals[threadIdx.x] = values[n1];
+  scales[rowIdx % MAX_BLOCK_THREADS] = r[rowIdx];
+
+  if(update) {
+    expnyts[rowIdx % MAX_BLOCK_THREADS] = std::exp(-labels[rowIdx] * losses[rowIdx]);
+    kappas[rowIdx % MAX_BLOCK_THREADS] = -labels[rowIdx] * expnyts[rowIdx % MAX_BLOCK_THREADS] / (1 + expnyts[rowIdx % MAX_BLOCK_THREADS]);
```
[the remark this answers is not in the document]
Comment (mdymczyk) [reply src/gpu/ffm/trainer.cu]: @henrygouk true and not sure - need to double check. For now I went with the original C implementation but need to experiment.

File: src/gpu/ffm/trainer.cu [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+      for (int d = 0; d < cK[0] * cWeightsOffset[0]; d+=cWeightsOffset[0]) {
+        T w1d = (weightsPtr + idx1)[d];
+        T w2d = (weightsPtr + idx2)[d];
+        const T g1 = cRegLambda[0] * w1d + kappas[rowIdx % MAX_BLOCK_THREADS] * w2d * v;
+        const T g2 = cRegLambda[0] * w2d + kappas[rowIdx % MAX_BLOCK_THREADS] * w1d * v;
+
+        const T w1gdup = (weightsPtr + idx1)[d+1] + g1 * g1;
+        const T w2gdup = (weightsPtr + idx2)[d+1] + g2 * g2;
+
+        (weightsPtr + idx1)[d] -= cLearningRate[0] / std::sqrt(w1gdup) * g1;
```
[the remark this answers is not in the document]
Comment (mdymczyk) [reply src/gpu/ffm/trainer.cu]: @henrygouk yes to both :-) 

1) definitely different updaters would be great, preferably abstracting this part and encapsulating it in an `Updater` class would be best. Can make an issue with potential candidates

2) for initial release I think it should be ok but in the long run I'm pretty sure indeterministic results will be a no-go, especially for our own use in DAI. Any suggestions how this could be fixed? Serializing it would incur a huge performance hit. Stochastic gradient descent? Not sure how viable that method is? Anything else?

File: src/gpu/ffm/trainer.cu [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+  scales[rowIdx % MAX_BLOCK_THREADS] = r[rowIdx];
+
+  if(update) {
+    expnyts[rowIdx % MAX_BLOCK_THREADS] = std::exp(-labels[rowIdx] * losses[rowIdx]);
+    kappas[rowIdx % MAX_BLOCK_THREADS] = -labels[rowIdx] * expnyts[rowIdx % MAX_BLOCK_THREADS] / (1 + expnyts[rowIdx % MAX_BLOCK_THREADS]);
+  }
+
+  __syncthreads();
+
+  T loss = 0.0;
```
[the remark this answers is not in the document]
Comment (mdymczyk) [reply src/gpu/ffm/trainer.cu]: In the original I think they call it `t`, not sure why - can rename. From my understanding, it is used as loss and also used to calculate predictions.

File: src/gpu/ffm/trainer.cu [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+  for(int i = 1; n1 + i < rowSizes[rowIdx + 1] - cBatchOffset[0]; i++) {
```
[the remark this answers is not in the document]
Comment (mdymczyk) [reply src/gpu/ffm/trainer.cu]: Not sure, instead of spinning up a thread for each field:feature:value (lets call it a node) tuple and running this loop for each consecutive node in that row we could spin up a thread for each node pair. Didn't try that approach yet as I was afraid of thread overhead.

Currently the main slowdown I'm noticing is becuase of:

1) number of registers used by the `wTx` kernel which limits the number of concurrent blocks being ran

2) read/write from global memory when using the weights array since that access isn't very coalesce within blocks. Interesting experiment is to move lines 157/158:

```
(weightsPtr + idx1)[d+1] = w1gdup;
(weightsPtr + idx2)[d+1] = w2gdup;
```

Right after:

```
const T w1gdup = (weightsPtr + idx1)[d+1] + g1 * g1;
const T w2gdup = (weightsPtr + idx2)[d+1] + g2 * g2;
```

There's a visible (~10%) slowdown on relatively large data (400k rows, 39 nodes in each row). I'm assuming this is due to how CUDA loads/stores data.

Basically the major slowdown in coming from the `if (update) {` branch. Putting both weights and gradients into the same array helped quite a bit but even on 1080Ti this is only ~2-3x faster than the CPU implementation.

**Comment by mdymczyk** [comment]
So both CPU and GPU implementations are there and working, the only issue left is that GPU batch mode gives slightly different results with same # of iterations (or converges in a much larger number of iterations) compared to GPU batch mode with batch_size=1 and CPU modes. I'm guessing this is because we are using HOGWILD! and the order of computations during gradient update differs (and might not be 100% correct?).

**Comment by mdymczyk** [comment]
One more thing: this needs to be compared against bigger data (libffm_toy.zip) and the original cpp implementation (https://github.com/guestwalk/libffm - not the Python API). I think the GPU version was getting a bit different results, so needs double checking before merging.
